### PR TITLE
feat: Integrate discovery and identify protocol

### PR DESCRIPTION
**Pull request by TheWaWaR**
## Changes
- [x] Add discovery protocol
- [x] Add identify protocol
- [x] Fix out bounded connect service
- [x] Fix can not shutdown network bug

## Known issues
- Shutdown network not very graceful.
- Identify protocol is just workaround, will update in later PR.

### CODE REVIEW

File: network/src/service/discovery_service.rs [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+        debug!(
+            "protocol [discovery] open on session [{}], address: [{}], type: [{:?}]",
+            session.id, session.address, session.ty
+        );
+        let event = DiscoveryEvent::Connected {
+            session_id: session.id,
+            peer_id: session.remote_pubkey.clone().map(|pubkey| pubkey.peer_id()),
+        };
+        // TODO: handle send failed
+        self.event_sender
```
Comment (jjyr) [inline network/src/service/discovery_service.rs]: ignore the error, the channel may be dropped when progress shutdown.

File: network/src/service/discovery_service.rs [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+        &mut self,
+        _control: &mut ServiceContext,
+        session: &SessionContext,
+        data: bytes::Bytes,
+    ) {
+        debug!("[received message]: length={}", data.len());
+
+        if let Some(ref mut sender) = self.discovery_senders.get_mut(&session.id) {
+            // TODO: handle channel is full (wait for poll API?)
+            if let Err(err) = sender.try_send(data.to_vec()) {
```
Comment (jjyr) [inline network/src/service/discovery_service.rs]: Send Bytes directly?

File: network/src/service/discovery_service.rs [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+        let (sender, receiver) = oneshot::channel();
+        // FIXME: what if send failed
+        let event = DiscoveryEvent::Misbehave {
+            session_id,
+            kind,
+            result: sender,
+        };
+        self.event_sender.unbounded_send(event).unwrap();
+        // FIXME: what if receive failed
+        receiver.wait().unwrap()
```
Comment (jjyr) [inline network/src/service/discovery_service.rs]: I think we can arbitrarily disconnect here.

File: network/src/protocol.rs [hunk 4]
```diff
@@ -92,8 +98,8 @@ impl ServiceProtocol for CKBHandler {
     fn init(&mut self, _control: &mut ServiceContext) {}
     fn connected(&mut self, control: &mut ServiceContext, session: &SessionContext, version: &str) {
         let (peer_id, version) = {
-            // FIXME: version number should be discussed.
-            let parsed_version = version.split('.').last().and_then(|v| v.parse::<u8>().ok());
+            // TODO: version number should be discussed.
+            let parsed_version = version.parse::<u8>().ok();
```
Comment (zhangsoledad) [inline network/src/protocol.rs]: How about use https://github.com/steveklabnik/semver ?

Comment (TheWaWaR) [reply network/src/protocol.rs]: Make sense to me, but need some API from underlying p2p library for select version.

File: network/src/service/discovery_service.rs [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+            }) => {
+                if let Some(peer_id) = peer_id {
+                    self.sessions.insert(session_id, peer_id);
+                }
+            }
+            Some(DiscoveryEvent::Disconnected(session_id)) => {
+                self.sessions.remove(&session_id);
+            }
+            Some(DiscoveryEvent::AddNewAddr { .. }) => {
+                // NOTE: ignore add new addr message, handle this in identify protocol
```
Comment (jjyr) [inline network/src/service/discovery_service.rs]: This event is confused, what's the different with AddNewAddrs?

File: network/src/service/discovery_service.rs [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+                        }
+                    }
+                }
+            }
+            Some(DiscoveryEvent::Misbehave {
+                session_id: _session_id,
+                kind: _kind,
+                result: _result,
+            }) => {
+                // FIXME:
```
Comment (jjyr) [inline network/src/service/discovery_service.rs]: Use `peer_store.ban_peer` for now

File: network/src/service/discovery_service.rs [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+    }
+
+    fn get_random(&mut self, n: usize) -> Vec<Multiaddr> {
+        let (sender, receiver) = oneshot::channel();
+        let event = DiscoveryEvent::GetRandom { n, result: sender };
+        if self.event_sender.unbounded_send(event).is_err() {
+            warn!(target: "network", "receiver maybe dropped!");
+            Vec::new()
+        } else {
+            receiver.wait().ok().unwrap_or_else(Vec::new)
```
Comment (jjyr) [inline network/src/service/discovery_service.rs]:
```suggestion
            receiver.wait().unwrap_or_else(Vec::new)
```

File: network/src/service/identify_service.rs [hunk 8]
```diff
@@ -1,236 +1,151 @@
...
-                        peer_id
-                        );
-                    }
+impl Stream for IdentifyService {
+    type Item = ();
+    type Error = ();
+    fn poll(&mut self) -> Result<Async<Option<Self::Item>>, Self::Error> {
+        match try_ready!(self.event_receiver.poll()) {
+            Some(IdentifyEvent::AddListenAddrs { peer_id, addrs }) => {
+                self.listen_addrs.insert(peer_id, addrs);
```
Comment (jjyr) [inline network/src/service/identify_service.rs]: should store addrs to peer_store if these addrs belongs to a peer.

File: network/src/service/identify_service.rs [hunk 9]
```diff
@@ -1,236 +1,151 @@
...
-                        "error when prepare identify : can't find addresses for peer {:?}",
-                        peer_id
-                        );
-                    }
+impl Stream for IdentifyService {
+    type Item = ();
+    type Error = ();
+    fn poll(&mut self) -> Result<Async<Option<Self::Item>>, Self::Error> {
+        match try_ready!(self.event_receiver.poll()) {
+            Some(IdentifyEvent::AddListenAddrs { peer_id, addrs }) => {
```
Comment (jjyr) [inline network/src/service/identify_service.rs]: `AddListenAddrs` -> `AddPeerAddrs` ?

File: network/src/service/outbound_peer_service.rs [hunk 10]
```diff
@@ -37,20 +37,20 @@ impl Stream for OutboundPeerService {
...
                         .read()
-                        .peers_to_attempt(new_outbound as u32);
-                    for (peer_id, addr) in attempt_peers.iter().filter_map(|(peer_id, addr)| {
-                        if self.network.local_peer_id() != peer_id {
-                            Some((peer_id.clone(), addr.clone()))
-                        } else {
-                            None
-                        }
-                    }) {
+                        .random_peers(new_outbound as u32)
```
Comment (jjyr) [inline network/src/service/outbound_peer_service.rs]: Pls revert to old logic

Comment (jjyr) [reply network/src/service/outbound_peer_service.rs]: The `attempt_peers` already includes filter logic

File: network/src/protocol.rs (hunk 4 above)
Comment (jjyr) [reply network/src/protocol.rs]: It is over enginery, network protocol should not be such complex.

File: network/src/service/discovery_service.rs (hunk 6 above)
Comment (TheWaWaR) [reply network/src/service/discovery_service.rs]: Better leave it here for now.

Comment (TheWaWaR) [reply network/src/service/discovery_service.rs]: Better leave it here for now.

File: network/src/service/identify_service.rs (hunk 8 above)
Comment (TheWaWaR) [reply network/src/service/identify_service.rs]: Refactor identify protocol later

File: network/src/service/discovery_service.rs (hunk 7 above)
Comment (TheWaWaR) [reply network/src/service/discovery_service.rs]: `Result::unwrap_or_else(|_| Vec::new)`
meaningless.

**Comment by jjyr** [comment]
There're still few problems in Identify protocol, should fix it after merge.

File: network/src/protocol.rs (hunk 4 above)
Comment (TheWaWaR) [reply network/src/protocol.rs]: Version select **is** [complex](https://github.com/bitcoin/bitcoin/blob/717fd58c4ba5c64778fa9a8bf6bbf5ae1164df3f/src/net_processing.cpp#L1808-L1936).
It's about you handle it in library or in user defined protocol.

**Pull request merged**